File: lib/domain/agricultural_task/interactors/agricultural_task_destroy_interactor.rb
# frozen_string_literal: true

module Domain
  module AgriculturalTask
    module Interactors
      class AgriculturalTaskDestroyInteractor < Domain::AgriculturalTask::Ports::AgriculturalTaskDestroyInputPort
        def initialize(output_port:, user_id:, gateway:, translator:, user_lookup:)
          @output_port = output_port
          @gateway = gateway
          @user_id = user_id
          @translator = translator
          @user_lookup = user_lookup
        end

        def call(task_id)
          user = @user_lookup.find(@user_id)
          access_filter = Domain::Shared::Policies::AgriculturalTaskPolicy.record_access_filter(user)
          result = @gateway.soft_delete_with_undo(
            user: user,
            task_id: task_id,
            auto_hide_after: 5000,
            translator: @translator,
            access_filter: access_filter
          )
          if result[:success]
            destroy_output_dto = Domain::AgriculturalTask::Dtos::AgriculturalTaskDestroyOutput.new(undo: result[:undo_entity])
            @output_port.on_success(destroy_output_dto)
          else
            @output_port.on_failure(result[:error_dto])
          end
        rescue Domain::Shared::Policies::PolicyPermissionDenied => e
          @output_port.on_failure(e)
        rescue Domain::Shared::Exceptions::RecordNotFound
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(@translator.t("agricultural_tasks.flash.not_found")))
        rescue Domain::Shared::Exceptions::RecordInvalid => e
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        end
      end
    end
  end
end

File: crates/agrr-domain/src/agricultural_task/interactors/agricultural_task_destroy_interactor.rs
//! Ruby: `Domain::AgriculturalTask::Interactors::AgriculturalTaskDestroyInteractor`

use crate::agricultural_task::dtos::AgriculturalTaskDestroyOutput;
use crate::agricultural_task::gateways::{AgriculturalTaskGateway, SoftDeleteUndoResult};
use crate::agricultural_task::ports::{AgriculturalTaskDestroyOutputPort, DestroyFailure};
use crate::shared::dtos::Error;
use crate::shared::exceptions::RecordNotFoundError;
use crate::shared::gateways::UserLookupGateway;
use crate::shared::policies::agricultural_task_policy;
use crate::shared::ports::translator_port::{TranslateOptions, TranslatorPort};
use crate::shared::reference_record_authorization;

pub struct AgriculturalTaskDestroyInteractor<'a, G, O, L, T> {
    output_port: &'a mut O,
    gateway: &'a G,
    user_id: i64,
    translator: &'a T,
    user_lookup: &'a L,
}

impl<'a, G, O, L, T> AgriculturalTaskDestroyInteractor<'a, G, O, L, T>
where
    G: AgriculturalTaskGateway,
    O: AgriculturalTaskDestroyOutputPort,
    L: UserLookupGateway,
    T: TranslatorPort,
{
    pub fn new(
        output_port: &'a mut O,
        user_id: i64,
        gateway: &'a G,
        translator: &'a T,
        user_lookup: &'a L,
    ) -> Self {
        Self { output_port, gateway, user_id, translator, user_lookup }
    }

    pub fn call(&mut self, task_id: i64) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let user = self.user_lookup.find(self.user_id);
        let access_filter = agricultural_task_policy::record_access_filter(user);
        let current = match self.gateway.find_by_id(task_id) {
            Ok(c) => c,
            Err(err) => {
                if err.downcast_ref::<RecordNotFoundError>().is_some() {
                    let message = self.translator.t(
                        "agricultural_tasks.flash.not_found",
                        &TranslateOptions::default(),
                    );
                    self.output_port.on_failure(DestroyFailure::Error(Error::new(message)));
                    return Ok(());
                }
                return Err(err);
            }
        };
        if let Err(policy) =
            reference_record_authorization::assert_edit_allowed(&access_filter, &current)
        {
            self.output_port.on_failure(DestroyFailure::Policy(policy));
            return Ok(());
        }
        let mut toast_opts = TranslateOptions::new();
        toast_opts.insert("name".into(), current.name.clone());
        let toast = self
            .translator
            .t("agricultural_tasks.undo.toast", &toast_opts);
        match self.gateway.soft_delete_with_undo(&user, task_id, 5000, &toast)? {
            SoftDeleteUndoResult::Success { undo } => {
                self.output_port
                    .on_success(AgriculturalTaskDestroyOutput::new(undo));
                Ok(())
            }
            SoftDeleteUndoResult::Failure { error } => {
                self.output_port.on_failure(DestroyFailure::Error(error));
                Ok(())
            }
        }
    }
}

#[cfg(test)]
mod interactors_agricultural_task_destroy_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/agricultural_task/interactors_agricultural_task_destroy_interactor_test.rs"));
}
